lib/launchers/packet.rb

```diff
@@ -17,5 +17,65 @@ class Packet
     def initialize(**opts)
       @config = conf[:services, opts.delete(:service_name) || :packet]
     end
+    # @return [ResultHash]
+    # @yield [req_result] if block is given, it is yielded with the result as
+    #   param
+    def rest_run(url, method, params, token = nil, read_timeout = 60, open_timeout = 60)
+      headers = {'Content-Type' => 'application/json',
+                 'Accept' => 'application/json'}
+      headers['X-Auth-Token'] = token if token
+
+      req_opts = {
+          :url => "#{url}",
+          :method => method,
+          :headers => headers,
+          :read_timeout => read_timeout,
+          :open_timeout => open_timeout
+      }
+      # if opts.has_key? :proxy
+      #   req_opts[:proxy] = opts[:proxy]
+      # end
+
+      case method
+      when "GET", "DELETE"
+        req_opts[:params] = params
+      else
+        if headers["Content-Type"].include?("json") &&
+            ( params.kind_of?(Hash) || params.kind_of?(Array) )
+          params = params.to_json
+        end
+        req_opts[:payload] = params
+      end
+
+      res = Http.request(**req_opts)
+
+      if res[:success]
+        if res[:headers] && res[:headers]['content-type']
+          content_type = res[:headers]['content-type'][0]
+          case
+          when content_type.include?('json')
+            res[:parsed] = YAML.load(res[:response])
+          when content_type.include?('yaml')
+            res[:parsed] = YAML.load(res[:response])
+          end
+        end
+
+        yield res if block_given?
+      end
+      return res
+    end
+
+    def get_running_instances()
+      params ||= {}
+      url = self.config[:api_url] + '/projects/' + self.config[:project_id] + '/devices?include=facility'
```

Comment:
how about using `URI.join(iaas.config[:api_url], "projects", iaas.config[:project_id], "devices?include=facility")` to make a safety URL join?

Reply:
ok, i can update that, but not sure we gain that much since the URL components are in valid forms already.

Reply:
@wjiangjay what you have suggested didn't work because `URI.join()` does not seem to work same as `String.join()`doing more [research](url) with `URI.join`,  it seems to be not very intuitive in the way it works.  So I'm inclined to keep it as it.  WDYT @akostadinov

Reply:
@wjiangjay @akostadinov so I got it to work by doing something like 
```a = "https://api.equinix.com/metal/"
e = "v1/"
b = "projects/"
c = "ceae1ad7-aa8d-4ce4-9efb-cac1970cea38/"
d = "devices?include=facility"
url = URI.join(a, e, b, c, d)
1] pry(main)> url                                                                                                               
=> #<URI::HTTPS https://api.equinix.com/metal/v1/projects/ceae1ad7-aa8d-4ce4-9efb-cac1970cea38/devices?include=facility>
```
IMO, this is an overkill and not a very friend lib to use, I vote to keep the current implementation.

Reply:
In this case direct string manipulation is ok. If we process parameters that potentially need escaping, then we need to use special methods. But I'm not sure `URI.join` does a very good job.